### PR TITLE
Add Erc1155 Code

### PR DESCRIPTION
**Pull request by liushooter**
Resolves # .

- @vivekvpandya 
- @LouiseMedova

### CODE REVIEW

File: erc1155/tests/tests.rs
Comment (LouiseMedova) [inline erc1155/tests/tests.rs:39]: According to the logic of a smart contract `operator` must be `USERS[1]`

File: erc1155/src/lib.rs
Comment (LouiseMedova) [inline erc1155/src/lib.rs:165]: Panic description is incorrect

File: erc1155/src/lib.rs
Comment (LouiseMedova) [inline erc1155/src/lib.rs:197]: The same: panic description is incorrect

File: erc1155/src/lib.rs
Comment (LouiseMedova) [inline erc1155/src/lib.rs:59]: What is this function for if there is already a function `get_balance`?

File: erc1155/src/lib.rs
Comment (LouiseMedova) [inline erc1155/src/lib.rs:148]: What is this function for if there is already a function `get_approval`?

File: erc1155/tests/tests.rs
Comment (LouiseMedova) [inline erc1155/tests/tests.rs:120]: Why codec?

File: erc1155/src/lib.rs
Comment (LouiseMedova) [reply erc1155/src/lib.rs:178]: It must be something like  "sender and recipient addresses are the same"

File: erc1155/src/lib.rs
Comment (LouiseMedova) [inline erc1155/src/lib.rs:176]: In that function you must check the account that sends that message (`msg::source`). It must the approved address or the owner.

Comment (LouiseMedova) [inline erc1155/src/lib.rs:176]: In that function you must check the account that sends that message (msg::source). It must the approved address or the owner.

File: erc1155/src/lib.rs
Comment (LouiseMedova) [inline erc1155/src/lib.rs:181]: That function is not finished

Comment (LouiseMedova) [inline erc1155/src/lib.rs:181]: It must fail if the the function `get_approval` returns false.
```suggestion
        if !*self.get_approval(from, to) {
```

File: erc1155/src/lib.rs
Comment (LouiseMedova) [inline erc1155/src/lib.rs:94]: I think it it better to name the variable `to` instead of `from
```suggestion
        if to == &ZERO_ID {
```

File: erc1155/src/lib.rs
Comment (LouiseMedova) [inline erc1155/src/lib.rs:121]: You must check in that function that account that sends the message (`msg::source()`). Otherwise anyone can burn anyone's tokens

File: erc1155/src/lib.rs
Comment (LouiseMedova) [inline erc1155/src/lib.rs:213]:
```suggestion
        if !*self.get_approval(from, to) {
```

File: erc1155/tests/tests.rs
Comment (LouiseMedova) [inline erc1155/tests/tests.rs:148]: Please add tests for checking panics

**Comment by LouiseMedova** [comment]
Since the functionality of the `mint` and `mint_batch` functions is the same, I suggest to make one function for minting tokens ( the same goes for `burn` and `burn_batch` functions). So the  function will receive `ids: &[u128], amounts: &[u128]` and in case of minting only one token these arrays will contain only one element (i.e.`mint` function), otherwise it will be `mint_batch` function.

File: erc1155/src/lib.rs
Comment (liushooter) [reply erc1155/src/lib.rs:164]: I rename get_approval to _get_approval,  treat it as an internal function.
the `is_approved_for_all` is the public function.

File: erc1155/src/lib.rs
Comment (liushooter) [reply erc1155/src/lib.rs:66]: I rename get_balance to _get_balance, treat it as an internal function.
the balance_of is the public function.

**Comment by liushooter** [comment]
> Since the functionality of the `mint` and `mint_batch` functions is the same, I suggest to make one function for minting tokens ( the same goes for `burn` and `burn_batch` functions). So the function will receive `ids: &[u128], amounts: &[u128]` and in case of minting only one token these arrays will contain only one element (i.e.`mint` function), otherwise it will be `mint_batch` function.

yeah，I think you make a good point. I reference the [code](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/token/ERC1155/ERC1155.sol#L266) .

I created the mint function for convenience. I recommend keeping this feature for the time being

**Comment by vivekvpandya** [comment]
should we need a test where transfer is made after approval is done.

File: erc1155/src/lib.rs
Comment (LouiseMedova) [inline erc1155/src/lib.rs:159]: What is the purpose of that function? It just calls `_get_approval`

File: erc1155/readme.md
Comment (LouiseMedova) [inline erc1155/readme.md:114]: Where that event is used?

**Comment by liushooter** [comment]
> should we need a test where transfer is made after approval is done.

yeah， i push the code https://github.com/gear-tech/apps/commit/a150a4bcfd9a5659275418c388c5d01f8dc6c660

File: erc1155/src/base.rs
Comment (LouiseMedova) [inline erc1155/src/base.rs:22]: That function isn't realized

File: erc1155/src/lib.rs
Comment (LouiseMedova) [inline erc1155/src/lib.rs:23]: You must add the mapping from token ids to the owner

File: erc1155/src/lib.rs
Comment (LouiseMedova) [inline erc1155/src/lib.rs:117]: Get the owner of `ids` from the mapping  and check that that owner is `msg_source()`

**Review by shamilsan — CHANGES_REQUESTED** [review]
Update to the new `gstd`

**Comment by liushooter** [comment]
> Update to the new `gstd`

ok

File: erc1155/src/lib.rs
Comment (liushooter) [reply erc1155/src/lib.rs:140]: fixed https://github.com/gear-tech/apps/commit/af0f12634f723d208f7de088ef4bacb910a139e0

File: erc1155/src/lib.rs
Comment (liushooter) [reply erc1155/src/lib.rs:23]: fixed https://github.com/gear-tech/apps/commit/af0f12634f723d208f7de088ef4bacb910a139e0

File: erc1155/src/base.rs
Comment (liushooter) [reply erc1155/src/base.rs:22]: fixed https://github.com/gear-tech/apps/commit/af0f12634f723d208f7de088ef4bacb910a139e0

**Comment by LouiseMedova** [comment]
Still don't see much sense in function `set_balance` and `get_balance` which contains only one line (if to remove `debug`). Why are there 2 functions for minting tokens and one function for burning?

**Pull request merged**
